# Benemérita Universidad Autónoma de Puebla
# Facultad de Ciencias de la Computación  
# Ingeniería en Ciencia de Datos  
#Materia: Introducción a la Ciencia de Datos  
#Profesor: Jaime Alejandro Romero Sierra  
#Alumno: Juan Diego Carreón Arellio  
#Fecha: 20/10/2025  
#Proyecto: Limpieza y Preparación de Datos – Detección de Fraudes 

En esta práctica se trabaja con un dataset de transacciones bancarias de Kaggle, modificado por el profesor para simular datos sucios.

El objetivo es **limpiar y preparar los datos** para su posterior análisis exploratorio, aplicando las funciones básicas de `pandas` vistas en clase:  
`isnull`, `drop_duplicates`, `fillna`, `replace`, `to_numeric`, `info`, `describe`.

El dataset tiene **10 columnas y más de 11,000 registros**.

In [ ]:
import pandas as pd
df=pd.read_csv('df_sucio.csv')
df

step      type     amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0       36  CASH_OUT  449109.84   C630750768            NaN            0.00   
1      402   PAYMENT    8593.72   C960271657           0.00            0.00   
2       33   PAYMENT    9201.16  C1378359079       22962.00        13760.84   
3       18       NaN    1225.30  C1707719646           0.00            0.00   
4      235   PAYMENT    9820.29    C90105308       10836.00         1015.71   
...    ...       ...        ...          ...            ...             ...   
11586  179  CASH_OUT  363585.95   C822058959       26708.00            0.00   
11587   13   CASH_IN  472601.05          NaN     2545417.97      3018019.02   
11588  375   CASH_IN  249638.23  C1324949601     4200851.80      4450490.03   
11589  329   PAYMENT    2786.10   C361627027        5349.00         2562.90   
11590  182   CASH_IN  301732.17   C671892507           0.00       301732.17   

          nameDest  oldbalanceDest newbalanceDest  isFraud  
0              NaN        72855.78      521965.62      0.0  
1       M431752332            0.00            0.0      0.0  
2       M352072570             NaN            0.0      0.0  
3      M1913693032            0.00            NaN      0.0  
4      M1581304819            0.00            0.0      0.0  
...            ...             ...            ...      ...  
11586   C325567260       173808.62      537394.57      0.0  
11587  C1839337592      6482126.00     6009524.95      0.0  
11588  C2037309464       256762.13         7123.9      0.0  
11589    M79674796            0.00            0.0      0.0  
11590    C60713113            0.00            0.0      0.0  

[11591 rows x 10 columns]

In [ ]:
df.head()

step      type     amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0   36  CASH_OUT  449109.84   C630750768            NaN            0.00   
1  402   PAYMENT    8593.72   C960271657            0.0            0.00   
2   33   PAYMENT    9201.16  C1378359079        22962.0        13760.84   
3   18       NaN    1225.30  C1707719646            0.0            0.00   
4  235   PAYMENT    9820.29    C90105308        10836.0         1015.71   

      nameDest  oldbalanceDest newbalanceDest  isFraud  
0          NaN        72855.78      521965.62      0.0  
1   M431752332            0.00            0.0      0.0  
2   M352072570             NaN            0.0      0.0  
3  M1913693032            0.00            NaN      0.0  
4  M1581304819            0.00            0.0      0.0

El comando `df.head()` muestra las primeras 5 filas del dataset.
Esto permite inspeccionar visualmente las columnas y detectar valores ausentes o incoherentes.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11591 entries, 0 to 11590
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   step            11128 non-null  object 
 1   type            11128 non-null  object 
 2   amount          11128 non-null  float64
 3   nameOrig        11128 non-null  object 
 4   oldbalanceOrg   11128 non-null  float64
 5   newbalanceOrig  11128 non-null  float64
 6   nameDest        11128 non-null  object 
 7   oldbalanceDest  11128 non-null  float64
 8   newbalanceDest  11128 non-null  object 
 9   isFraud         10910 non-null  float64
dtypes: float64(5), object(5)
memory usage: 905.7+ KB


In [ ]:
df.isna().sum()

step              463
type              463
amount            463
nameOrig          463
oldbalanceOrg     463
newbalanceOrig    463
nameDest          463
oldbalanceDest    463
newbalanceDest    463
isFraud           681
dtype: int64

In [ ]:
df.duplicated().sum()

np.int64(577)

Con `df.info()` verificamos los tipos de datos y el número de valores no nulos.  
`df.isna().sum()` permite identificar cuántos valores faltan por columna.  
`df.duplicated().sum()` cuenta cuántas filas están repetidas.

In [ ]:
# === 2. ELIMINACIÓN DE DUPLICADOS ===
antes = df.shape[0]
df = df.drop_duplicates()
despues = df.shape[0]

print(f"Duplicados eliminados: {antes - despues}")


Duplicados eliminados: 577


Se eliminaron 577 registros duplicados.  
Esto evita sesgos en los cálculos estadísticos o el entrenamiento de modelos.


In [ ]:
# === 3. CONVERSIÓN DE TIPOS ===
cols_num = ["step", "amount", "oldbalanceOrg", "newbalanceOrig",
            "oldbalanceDest", "newbalanceDest", "isFraud"]
cols_cat = ["type", "nameOrig", "nameDest"]

# Convertir columnas numéricas
for col in cols_num:
    if col in df.columns:
        df[col] = pd.to_numeric(df[col], errors="coerce")

# Convertir columnas categóricas a texto limpio
for col in cols_cat:
    if col in df.columns:
        df[col] = df[col].astype(str).str.strip().str.lower()


C:\Users\Uusario Nuevoo\AppData\Local\Temp\ipykernel_8408\885527680.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = pd.to_numeric(df[col], errors="coerce")
C:\Users\Uusario Nuevoo\AppData\Local\Temp\ipykernel_8408\885527680.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = df[col].astype(str).str.strip().str.lower()
C:\Users\Uusario Nuevoo\AppData\Local\Temp\ipykernel_8408\885527680.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.


Con `to_numeric(errors="coerce")` se convierten valores incorrectos a `NaN`, 
y luego se corrigen.  
Las variables categóricas se normalizan (sin mayúsculas ni espacios extra).


In [ ]:
# === 4. ANÁLISIS DE VALORES FALTANTES ===
faltantes = (df.isna().sum() / len(df) * 100).round(2)
faltantes


step              6.21
type              0.00
amount            4.19
nameOrig          0.00
oldbalanceOrg     4.19
newbalanceOrig    4.19
nameDest          0.00
oldbalanceDest    4.19
newbalanceDest    6.20
isFraud           6.16
dtype: float64

Este paso calcula el **porcentaje de valores faltantes** por columna.  
Nos permite priorizar qué columnas requieren imputación.


In [ ]:
# === 5. RELLENO DE VALORES NULOS ===
for col in cols_num:
    if df[col].isna().sum() > 0 and col != "isFraud":
        df[col] = df[col].fillna(df[col].median())

if "isFraud" in df.columns:
    df["isFraud"] = df["isFraud"].fillna(0)

for col in cols_cat:
    if df[col].isna().sum() > 0:
        df[col] = df[col].fillna("desconocido")


C:\Users\Uusario Nuevoo\AppData\Local\Temp\ipykernel_8408\3403717811.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = df[col].fillna(df[col].median())
C:\Users\Uusario Nuevoo\AppData\Local\Temp\ipykernel_8408\3403717811.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = df[col].fillna(df[col].median())
C:\Users\Uusario Nuevoo\AppData\Local\Temp\ipykernel_8408\3403717811.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[ro

- Se rellenan los valores faltantes numéricos con la **mediana**, 
  ya que es robusta ante outliers.  
- Las variables categóricas se completan con `"desconocido"`.  
- La columna `isFraud` se rellena con **0**, indicando “no fraude”.


In [ ]:
# === 6. DETECCIÓN DE OUTLIERS POR IQR ===
for col in ["amount", "oldbalanceOrg", "oldbalanceDest"]:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    low = Q1 - 1.5 * IQR
    high = Q3 + 1.5 * IQR
    n_out = df[(df[col] < low) | (df[col] > high)].shape[0]
    print(f"{col}: {n_out} outliers detectados")


amount: 599 outliers detectados
oldbalanceOrg: 1861 outliers detectados
oldbalanceDest: 1473 outliers detectados


El método **IQR (Interquartile Range)** identifica posibles valores atípicos 
en columnas numéricas.  
No se eliminan, solo se registran, como pidió el profesor.


In [ ]:
# === 7. VERIFICACIÓN FINAL ===
print(df.info())
print(df.isna().sum())


<class 'pandas.core.frame.DataFrame'>
Index: 11014 entries, 0 to 11589
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   step            11014 non-null  float64
 1   type            11014 non-null  object 
 2   amount          11014 non-null  float64
 3   nameOrig        11014 non-null  object 
 4   oldbalanceOrg   11014 non-null  float64
 5   newbalanceOrig  11014 non-null  float64
 6   nameDest        11014 non-null  object 
 7   oldbalanceDest  11014 non-null  float64
 8   newbalanceDest  11014 non-null  float64
 9   isFraud         11014 non-null  float64
dtypes: float64(7), object(3)
memory usage: 946.5+ KB
None
step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
dtype: int64


Al final del proceso, `info()` muestra que ya no existen valores nulos, 
y todos los tipos de datos son correctos.  
Esto confirma que la limpieza fue exitosa.


In [ ]:
# === 8. EXPORTACIÓN ===
df.to_csv("transacciones_limpias.csv", index=False)
print("Archivo limpio guardado como: transacciones_limpias.csv")


Archivo limpio guardado como: transacciones_limpias.csv



#CONCLUSIONES

1. Se eliminaron 577 duplicados y se corrigieron tipos de datos.  
2. Se reemplazaron valores faltantes con la mediana o "desconocido".  
3. Se identificaron outliers mediante IQR, sin eliminarlos.  
4. No se utilizó `dropna()`, cumpliendo con la consigna del profesor.  
5. La base final no presenta nulos ni duplicados, y está lista para análisis.  

---
**Dataset limpio final:** `transacciones_limpias.csv`
